lib/signs_ui/alerts/display.ex

```diff
@@ -0,0 +1,64 @@
+defmodule SignsUi.Alerts.Display do
+  alias SignsUi.Alerts.Alert
+  alias SignsUi.Alerts.State
+
+  @typep alert_map :: %{Alert.id() => Alert.t()}
+  @typep route_alerts_map :: %{Alert.route_id() => alert_map()}
+
+  defstruct alerts: %{}
+
+  @typep t :: %__MODULE__{
+          alerts: route_alerts_map()
+        }
+
+  @spec format_state(State.t()) :: t()
+  def format_state(state) do
+    %__MODULE__{
+      # Take every alert in the current state,
+      alerts:
+        state
+        # generate a list of single_route_alerts from it, and flatten,
+        |> Stream.flat_map(&expand_routes/1)
+        # group them by route,
+        |> Enum.group_by(fn alert -> alert.route end)
+        # convert each group's list of alerts into a map,
+        |> Stream.map(&alert_map_for_route/1)
+        # and convert the list of alert groups into a map.
+        |> Map.new()
```

Comment:
I think what you have here is understandable and nice. But never one to pass up a good code golfing sesh, I might take advantage of `Map.new/2`, which takes a mapping function, the `& &1` anonymous function syntax. That would leave something like:

```ex
state
|> Stream.flat_map(&expand_routes/1)
|> Enum.group_by(& &1.route)
|> Map.new(&alert_map_for_route/1)

def alert_map_for_route({route, alerts}) do
  {route, Map.new(alerts, &{&1.id, &1})}
end
```